zcash_primitives/CHANGELOG.md

```diff
@@ -131,6 +139,9 @@ and this library adheres to Rust's notion of
     - `fees::fixed::FeeRule::fixed_fee`
     - `fees::zip317::FeeRule::marginal_fee`
     - `sighash::TransparentAuthorizingContext::input_amounts`
+- `zcash_primitives::zip32`:
+  - `ChildIndex` has been changed from an enum to an opaque struct, and no
+    longer supports non-hardened indices.
```

Comment:
Do you want to add a note to ZIP 32 that discourages or deprecates use of non-hardened indices with Sapling?

I feel like this requires more community discussion. I'll raise it at the ZIP sync meeting today.

Reply:
ZIP 32's existing language already somewhat discourages it, particularly the language that specifies Orchard. We only specify non-hardened derivation for an optional `address_index` path extension for Sapling with MAY, and then immediately note that `zcashd` did this path extension with hardened derivation instead (which made sense for its use case, and also means I still am unaware of anyone ever exercising this MAY).

I also note that while we do still use a non-hardened-like derivation process to produce Sapling internal keys, it is domain-separated from non-hardened derivation and thus distinct (so we don't use the functionality being removed here).